eland/operations.py

```diff
@@ -1233,58 +1233,55 @@ def _es_results(
-        else:
-            body["_source"] = False
-
-        es_results: Any = None
-
-        # If size=None use scan not search - then post sort results when in df
-        # If size>10000 use scan
-        is_scan = False
-        if size is not None and size <= DEFAULT_ES_MAX_RESULT_WINDOW:
-            if size > 0:
-                es_results = query_compiler._client.search(
-                    index=query_compiler._index_pattern,
-                    size=size,
-                    sort=sort_params,
-                    body=body,
-                )
-        else:
-            is_scan = True
-            es_results = scan(
-                client=query_compiler._client,
-                index=query_compiler._index_pattern,
-                query=body,
-            )
-            # create post sort
-            if sort_params is not None:
-                post_processing.append(SortFieldAction(sort_params))
-
-        if is_scan:
-            while True:
-                partial_result, df = query_compiler._es_results_to_pandas(
-                    es_results, collector.batch_size(), collector.show_progress
-                )
-                df = self._apply_df_post_processing(df, post_processing)
-                collector.collect(df)
-                if not partial_result:
-                    break
+        body["_source"] = _source if _source else False
+
+        is_pagination_required: bool = False
+
+        if result_size and (0 <= result_size <= DEFAULT_ES_MAX_RESULT_WINDOW):
+            body["size"] = result_size
         else:
-            partial_result, df = query_compiler._es_results_to_pandas(es_results)
-            df = self._apply_df_post_processing(df, post_processing)
-            collector.collect(df)
+            body["size"] = DEFAULT_ES_MAX_RESULT_WINDOW
+            is_pagination_required = True
+
+        if sort_params is not None:
+            body["sort"] = [sort_params]
+        elif is_pagination_required:
+            body["sort"] = [{"_doc": "asc"}]
+
+        # To Improve performance
+        body["track_total_hits"] = False
+
+        # Initialize Point in time API
+        pit_id: str = query_compiler._client.open_point_in_time(
+            index=query_compiler._index_pattern, keep_alive="1m"
+        )["id"]
+        body["pit"] = {"id": pit_id, "keep_alive": "1m"}
+
+        es_results: List[Dict[str, Any]] = []
+
+        # Check result_size to handle df.head(0) and df.tail(0)
+        while True and result_size != 0:
+            results = query_compiler._client.search(body=body)["hits"]["hits"]
+            es_results.extend(results)
+            # If results is empty, we completed searching
+            if results and is_pagination_required:
+                body["search_after"] = results[-1]["sort"]
+            else:
+                break
+
+        try:
+            res = query_compiler._client.close_point_in_time(body={"id": pit_id})
+            assert res["succeeded"] is True
```

Comment:
No need to assert anything here, we tried our best and the PIT will expire after the keep_alive anyways.